dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionDataConsolidator.java

```diff
@@ -0,0 +1,260 @@
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.predictor;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import lombok.Setter;
+
+import org.hisp.dhis.analytics.AnalyticsService;
+import org.hisp.dhis.category.CategoryService;
+import org.hisp.dhis.common.DimensionalItemObject;
+import org.hisp.dhis.common.FoundDimensionItemValue;
+import org.hisp.dhis.dataelement.DataElement;
+import org.hisp.dhis.dataelement.DataElementOperand;
+import org.hisp.dhis.datavalue.DataValueService;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.period.Period;
+
+/**
+ * Consolidates the prediction data for one organisation unit at a time,
+ * combining both aggregate data values and analytics data values.
+ *
+ * @author Jim Grace
+ */
+public class PredictionDataConsolidator
+{
+    @Setter // to Mock for testing
+    private PredictionDataValueFetcher dataValueFetcher;
+
+    @Setter // to Mock for testing
+    private PredictionAnalyticsDataFetcher analyticsFetcher;
+
+    private Set<Period> dataValueQueryPeriods;
+
+    private Set<Period> analyticsQueryPeriods;
+
+    private Set<Period> existingOutputPeriods;
+
+    private DataElementOperand outputDataElementOperand;
+
+    private Set<DataElement> dataElements;
+
+    private Set<DataElementOperand> dataElementOperands;
+
+    private Set<DimensionalItemObject> analyticsItems;
+
+    private Queue<OrganisationUnit> orgUnitsRemaining;
+
+    private Queue<PredictionData> readyPredictionData;
```

Comment:
I take it the goal would be to have as few non-final fields as possible? To this end, I've eliminated the `setUp` method which was admittedly a bit awkward to begin with, as it was only called once. I moved two of its arguments to the constructor, where they can be processed and set into `final` fields or just passed on to other classes, and moved the other arguments to `init` where they can directly feed the two fetcher `init` methods.

The reasons I had `setUp` in the first place were (a) it allowed the constructor to have only services for arguments, following the pattern of most of our service constructors, and (b) it allowed the calls within `DefaultPredictionService` to `consolidator.init` to pass only the values that were different from one `OrganisationUnitLevel` to the next, and not the parameters that were unchanging throughout the whole predictor. I believe it made this method more clear. But I have refactored it now because (a) the reason for patterning the constructor after services was not a strong one, especially since this class is always constructed by hand, and (b) the benefit of fewer arguments to `consolidator.init` might be outweighed by the benefit of removing the `setUp` method altogether, as well as the benefit of having fewer mutable class fields in `PredictionDataConsolidator`.

Further, I have moved the construction of `dataValueFetcher` and `analyticsFetcher` to the caller of the `PredictionDataConsolidator` constructor. That way they don't need to be mutable fields for mocking in a unit test; either they are constructed by `DefaultPredictionService` or they are mocked by `PredictionDataConsolidatorTest`. When writing the code I had gone back and forth about where these classes should be instantiated. At first I thought they should be instantiated by the caller to `PredictionDataConsolidator` and passed on construction, as they are now. But before I submitted the code I had them constructed by `PredictionDataConsolidator` because this hid what the consolidator needed from its caller. But I've now gone back to having the caller construct them because it further reduces the mutable fields of `PredictionDataConsolidator`.

As a result, the only mutable fields of `PredictionDataConsolidator` are the `Queues` `orgUnitsRemaining`, and `readyPredictionData`, which really represent the state of the object, plus `analyticsBatchFetchSize` just because it can be overridden for unit testing. I agree that this distinction between mutable and immutable makes it much clearer what can happen in the class.

Reply:
> I take it the goal would be to have as few non-final fields as possible?

Generally, yes :D - It usually makes code and the process it represents much easier to reason about.